internal/controller/istiogatewaysecret/setup.go

```diff
@@ -0,0 +1,81 @@
+package istiogatewaysecret
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	apicorev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlruntime "sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/kyma-project/lifecycle-manager/api/shared"
+	"github.com/kyma-project/lifecycle-manager/internal/pkg/flags"
+	"github.com/kyma-project/lifecycle-manager/pkg/gatewaysecret"
+	gatewaysecretclient "github.com/kyma-project/lifecycle-manager/pkg/gatewaysecret/client"
+)
+
+const (
+	controllerName    = "istio-controller"
+	kcpRootSecretName = "klm-watcher"
+)
+
+var errCouldNotGetLastModifiedAt = errors.New("getting lastModifiedAt time failed")
+
+func SetupReconciler(mgr ctrl.Manager, flagVar *flags.FlagVar, options ctrlruntime.Options) error {
+	options.MaxConcurrentReconciles = flagVar.MaxConcurrentKymaReconciles
+
+	clnt := gatewaysecretclient.NewGatewaySecretRotationClient(mgr.GetConfig())
+	var parseLastModifiedFunc gatewaysecret.TimeParserFunc = func(secret *apicorev1.Secret) (time.Time, error) {
+		if gwSecretLastModifiedAtValue, ok := secret.Annotations[shared.LastModifiedAtAnnotation]; ok {
+			if gwSecretLastModifiedAt, err := time.Parse(time.RFC3339, gwSecretLastModifiedAtValue); err == nil {
+				return gwSecretLastModifiedAt, nil
+			}
+		}
+		return time.Time{}, errCouldNotGetLastModifiedAt
+	}
+	handler := gatewaysecret.NewGatewaySecretHandler(clnt, parseLastModifiedFunc)
+
+	var getSecretFunc GetterFunc = func(ctx context.Context, name types.NamespacedName) (*apicorev1.Secret, error) {
+		secret := &apicorev1.Secret{}
```

Comment:
so why are we not using the client here so that it is clnt.GetGatewaysecret? and then we only use the request namespaced name to check first if it is the actual rootsecret name and namespace since anyways it should be them because we have the secret predicate? I mean we shouldn't be fetching the secret if the name or namespace are not those of the root secret in the reconciler if it somehow enters the reconciliation loop despite of the predicates

Reply:
Handler and Reconciler are not sharing the same dependency and the `GetterFunc` here might also be obsolete, due to the investigation I plan considering if we can use this other Reconciler interface:
https://github.com/kubernetes-sigs/controller-runtime/blob/aea2e32a936584b06ae6f7992f856fe7292b0297/pkg/reconcile/reconcile.go#L131

For the sake of encapsulation I want this dependency to be separately tested in `controller_test` and the scope of the Reconciler type.

Reply:
I am not sure I understand this. This is just the setup of the reconciler so why not just fetch the secret that the reconciler is intended for? what I mean is something like this:
var getSecretFunc GetterFunc = func(ctx context.Context, name types.NamespacedName) (*apicorev1.Secret, error) {
		if name.Name != shared.GatewaySecretName || name.Namespace != shared.IstioNamespace {
			return nil, fmt.Errorf("invalid secret name %s or namespace %s", name.Name, name.Namespace)
		}

		secret, err := clnt.GetGatewaySecret(ctx)
		if err != nil {
			return nil, fmt.Errorf("failed to get root gateway secret %w", err)
		}

		return secret, nil
	}

Reply:
Because the root secret is different from gateway secret 👍